### PR TITLE
Update robust glm notebook

### PR DESCRIPTION
**Pull request by jonsedar**
The **first purpose of this PR** is to update the GLM robust regression notebook already in the examples here: https://docs.pymc.io/notebooks/GLM-robust-with-outlier-detection.html 

Those updates are everything from v2.1 onwards:

Version history:

version | date | author | changes
:--- | :--- | :--- | :---
1.0 | 2015-12-21 | [jonsedar](https://github.com/jonsedar) | Create and publish
2.0 | 2018-07-24 | [twiecki](https://github.com/twiecki) | Restate outlier model using `pm.Normal.dist().logp()` and `pm.Potential()`
2.1 | 2019-11-16 | [jonsedar](https://github.com/jonsedar) | Restate `nu` in StudentT model to be more efficient, drop explicit use of theano shared vars, generally improve plotting / explanations / layout
2.2 | 2020-05-21 | [jonsedar](https://github.com/jonsedar) | Minor tidyup for plots and warnings and rerun with pymc3.8

The **second purpose of this PR** is to clarify the docstring within `sampling.py`, specifically the kwargs for `step_kwargs` when you have multiple steppers. I found the need for better clarity in the docstring during my rework of the notebook, so I hope it's valid to include a change in this single PR. I think this is also a fix for https://github.com/pymc-devs/pymc3/issues/3197

### CODE REVIEW

**Comment by twiecki** [comment]
Wow, this is an amazing improvement on an already amazing post.

The one thing I want to make sure is that the headings are correct -- are section headings starting with `##` rather than `#` which is reserved only for the title? Otherwise this creates problems with TOCs.

File: pymc3/sampling.py
Comment (twiecki) [inline pymc3/sampling.py]: for nuts this also works if you pass it directly, right? This might make it sound like you always have to use this extra-nesting.

Comment (AlexAndorra) [reply pymc3/sampling.py]: Yes, it works if you pass it directly, i.e `pm.sample(target_accept=0.9)`. You have to pass the step method only when the sampler uses several step methods.
So, I agree with Thomas and think this clarification should be mentionned.

Comment (jonsedar) [reply pymc3/sampling.py]: Thanks for the feedback guys - I've clarified the language in my latest commit

**Comment by jonsedar** [comment]
Thanks for the feedback guys - I've tidied the notebook and clarified the docstring language in my latest commit

**Review by twiecki — APPROVED** [review]

**Comment by AlexAndorra** [comment]
Thanks Jon! I'm reviewing the NB right now -- will let you know when I'm done

**Comment by AlexAndorra** [comment]
Thanks Jon, this looks really nice and is a clear improvement over the old NB! I posted the first part of my review of your NB above, and I'll review the second part later this afternoon 😉

**Comment by jonsedar** [comment]
Thanks Alex, I'll wait for you to complete the review before I make any changes that you've suggested.

Cheers,

**Comment by AlexAndorra** [comment]
Yeah that'd be best if you can. I'm almost finished.

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
The docstring looks good now 👌 I had a few comments for the NB -- should be ready to merge after that.
This is a great update, thanks Jon!

**Comment by jonsedar** [comment]
Righto - thanks for the thorough review! I'll make changes after lunch :)

**Comment by jonsedar** [comment]
<p>Yep - removed. The important bits (python 3.6, pymc3 3.8) are already noted</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>probably worth keeping incase people run this in other IDEs (VSCode, PyCharm, Syder etc) I'm not sure how those work</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Sure, done.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Yeah, that can go - I tend to keep this around in my WIP templates as a location to abstract away long functions before they go to a source file</p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>ta :) though sometimes I really miss ggplot text_geom</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>good catch</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Cool, will make a mental note to update this once that's out :)</p><p> </p><p>I kept the defaults in since I think it can be useful to see them</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Good to know - I'll definitely look out for that in future then: I find joint dists are massively helpful</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Gotcha :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Thanks, I've made a note to come back update arvin throughout when available https://github.com/jonsedar/pymc3_examples/issues/15</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>good catch</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Fair point, added a note:</p><p>testval for <code>is_outlier</code> initialised in order to create class asymmetry</p><p> </p><p>The other testvals <em>were</em> set also in order to create asymmetry, but I've found they don't need it. Have reverted to mean values for b0, b1, y_est_out, sigma_y_out</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>The passing of time eh? have updated to arviz=0.7 released between initial draft writing and now. I've been able to remove the warning suppression</p><p> </p><p>Will remove the <code>combined</code> and <code>compact</code>   kwargs: I don't think they used to be false by default. All good</p><p> </p><p>That's a nifty trick with the <code>var_names</code>,  will remember that for later. I think for now though it's safer to make explicit wouldn't you say?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Outliers ought to be the minority class (and might not even exist) in this modelspec, so I think it's fair to constrain them Uniform[0, 0.5], rather than Beta(0, 1)...</p><p> </p><p>But, that said, Uniform is a bit of a cludge. Could possibly use a distribution with a more central tendency, but would have to truncate it. Maybe there's another modelspec somewhere :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>changed to az </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>righto - I've logged a note to update these at some point :) https://github.com/jonsedar/pymc3_examples/issues/15</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>Added to the list :) https://github.com/jonsedar/pymc3_examples/issues/15</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
<p>ah sweet typos... good catch, thanks!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3908/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by jonsedar** [comment]
okie dokie, I've addressed all @AlexAndorra's points - great review thank you - and I think this is ready to go

One note: I did the olde:

```
❯ git fetch upstream
❯ git rebase upstream/master
```

prior to my latest commit in this PR and see quite a few files changed in the meantime , must be a weekend with lots of folks working :)

I don't see any clashes, but do let me know if you'd prefer a clean PR

Cheers, Jon

**Comment by fonnesbeck** [comment]
Hey, Jon. Looks great. Why is section 4.2.3 a markdown cell and not code?

**Comment by AlexAndorra** [comment]
Thanks for the updates Jon! I'll review that tomorrow

Le dim. 3 mai 2020 à 21:20, Jonathan Sedar <notifications@github.com> a
écrit :

**Comment by jonsedar** [comment]
> Hey, Jon. Looks great. Why is section 4.2.3 a markdown cell and not code?

Quite right - I'm an idiot! Will update now

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Thanks for the updates Jon, this all looks good now! Just one last thing (sorry, forgot to tell you about it yesterday): could you please Black-format the NB with [black_nbconvert](https://github.com/dfm/black_nbconvert)? 
That's what we use on the resources repo to standardize NBs: `pip install black_nbconvert`, then `black_nbconvert /path/to/a/notebook.ipynb`. 
After that I'll merge 🍾 

A lot of files have indeed been picked up by rebasing your branch. I don't know how you guys usually do, but I think it's no prb, as there is no conflicts and I'll squash & merge anyway.

**Comment by jonsedar** [comment]
Hi Alex,

Good call re: https://github.com/dfm/black_nbconvert, that's a new one to me, ands no surprise to see it's by the ever-helpful & prolific DanFM!

I've just run that reformatting and a re-run, all seems well, so will commit now

**Comment by jonsedar** [comment]
Interesting test failure, if this requires underlying fixes to pymc3 let me know and perhaps for simplicity I can just make a new clean PR

**Review by AlexAndorra — APPROVED** [review]
Thanks a lot Jon, this is all good now 👌 
I'll merge as soon as tests pass -- this error is super weird BTW. It looks like pip couldn't install numpy. Do you have an idea how to fix that?

**Comment by jonsedar** [comment]
No idea - it could just have been a network hiccup at the time... For good measure I'll recommit and trigger again!

**Comment by jonsedar** [comment]
It was a ghost in the machine :D

**Comment by AlexAndorra** [comment]
Apparently 😅 
Thanks again for this great update Jon!

**Comment by jonsedar** [comment]
Thanks for all your help and reviews! Will try to pick up some bugs next...

BTW is there a protocol for deleting this branch on origin?

**Comment by AlexAndorra** [comment]
Good question -- I'm not aware of any such protocol, but that'd be good practice

**Pull request merged**
